File: src/monkey/lexer/lexer.py
from monkey.tokens.tokens import Token, Tokens, lookup_ident


def is_letter(char):
    return isinstance(char, str) and ("a" <= char <= "z" or "A" <= char <= "Z" or char == "_")


def is_digit(char):
    return isinstance(char, str) and ("0" <= char <= "9")


class Lexer:
    """The lexer class takes an input string and provides a
    `next_token` function to be iterated over by the parser.

    `self.char` contains the most recently read character in the input string
    `self.position` is the index of that character in the input string
    `self.read_position` is always ahead of `self.position` and points to the next character to be read
    """

    def __init__(self, input: str) -> None:
        self.input = input
        self.read_position = 0
        self._read_char()

    def _read_char(self):
        """Returns the next character (ASCII) in `self.input."""
        if self.read_position >= len(self.input):
            self.char = None
        else:
            self.char = self.input[self.read_position]

        self.position = self.read_position
        self.read_position += 1

    def _read_identifier(self):
        init_pos = self.position
        while is_letter(self.char):
            self._read_char()

        return self.input[init_pos : self.position]

    def _read_number(self):
        init_pos = self.position
        while is_digit(self.char):
            self._read_char()

        return self.input[init_pos : self.position]

    def _skip_whitespace(self):
        while self.char in [" ", "\t", "\n", "\r"]:
            self._read_char()

    def _peek_char(self):
        if self.read_position >= len(self.input):
            return None
        else:
            return self.input[self.read_position]

    def next_token(self):
        tok: Token

        self._skip_whitespace()

        match self.char:
            case "=":
                if self._peek_char() == "=":
                    first_char = self.char
                    self._read_char()
                    tok = Token(type=Tokens.EQ, literal=first_char + self.char)
                else:
                    tok = Token(type=Tokens.ASSIGN, literal=self.char)
            case "+":
                tok = Token(type=Tokens.PLUS, literal=self.char)
            case "-":
                tok = Token(type=Tokens.MINUS, literal=self.char)
            case "!":
                if self._peek_char() == "=":
                    first_char = self.char
                    self._read_char()
                    tok = Token(type=Tokens.NOT_EQ, literal=first_char + self.char)
                else:
                    tok = Token(type=Tokens.BANG, literal=self.char)
            case "/":
                tok = Token(type=Tokens.SLASH, literal=self.char)
            case "*":
                tok = Token(type=Tokens.ASTERISK, literal=self.char)
            case "<":
                tok = Token(type=Tokens.LT, literal=self.char)
            case ">":
                tok = Token(type=Tokens.GT, literal=self.char)
            case "(":
                tok = Token(type=Tokens.LPAREN, literal=self.char)
            case ")":
                tok = Token(type=Tokens.RPAREN, literal=self.char)
            case "{":
                tok = Token(type=Tokens.LBRACE, literal=self.char)
            case "}":
                tok = Token(type=Tokens.RBRACE, literal=self.char)
            case ",":
                tok = Token(type=Tokens.COMMA, literal=self.char)
            case ";":
                tok = Token(type=Tokens.SEMICOLON, literal=self.char)
            case None:
                tok = Token(type=Tokens.EOF, literal="")
            case _:
                if is_letter(self.char):
                    literal = self._read_identifier()
                    tok = Token(
                        type=lookup_ident(literal),
                        literal=literal,
                    )
                    return tok
                elif is_digit(self.char):
                    tok = Token(type=Tokens.INT, literal=self._read_number())
                    return tok
                else:
                    tok = Token(type=Tokens.ILLEGAL, literal=self.char)

        self._read_char()
        return tok

File: rust/src/lexer.rs
use crate::token::Token;
use anyhow::Result;

pub struct Lexer {
    input: Vec<u8>,
    position: usize,
    read_position: usize,
    ch: u8,
}

impl Lexer {
    pub fn new(input: String) -> Lexer {
        let mut lexer = Lexer {
            input: input.as_bytes().to_vec(),
            position: 0,
            read_position: 0,
            ch: 0,
        };
        lexer.read_char();
        lexer
    }

    fn read_char(&mut self) {
        if self.input.len() <= self.read_position {
            self.ch = 0;
        } else {
            self.ch = self.input[self.read_position];
        }
        self.position = self.read_position;
        self.read_position += 1;
    }

    fn peek_char(&self) -> u8 {
        if self.input.len() <= self.read_position {
            0
        } else {
            self.input[self.read_position]
        }
    }

    fn skip_whitespace(&mut self) {
        while self.ch.is_ascii_whitespace() {
            self.read_char()
        }
    }

    pub fn next_token(&mut self) -> Result<Token> {
        self.skip_whitespace();
        let tok = match self.ch {
            b'+' => Token::Plus,
            b'-' => Token::Minus,
            b'(' => Token::LParen,
            b')' => Token::RParen,
            b'{' => Token::LBrace,
            b'}' => Token::RBrace,
            b'>' => Token::GreaterThan,
            b'<' => Token::LessThan,
            b',' => Token::Comma,
            b';' => Token::Semicolon,
            b'*' => Token::Asterisk,
            b'/' => Token::Slash,
            b'=' => {
                if self.peek_char() == b'=' {
                    self.read_char();
                    Token::Equal
                } else {
                    Token::Assign
                }
            }
            b'!' => {
                if self.peek_char() == b'=' {
                    self.read_char();
                    Token::NotEqual
                } else {
                    Token::Bang
                }
            }
            b'a'..=b'z' => {
                let ident = self.read_identifier();
                return Ok(match ident.as_str() {
                    "let" => Token::Let,
                    "fn" => Token::Function,
                    "if" => Token::If,
                    "else" => Token::Else,
                    "return" => Token::Return,
                    "true" => Token::True,
                    "false" => Token::False,
                    _ => Token::Ident(ident),
                });
            }
            b'0'..=b'9' => return Ok(Token::Int(self.read_number())),
            0 => Token::Eof,
            _ => Token::Illegal,
        };
        self.read_char();
        Ok(tok)
    }

    fn read_identifier(&mut self) -> String {
        let start = self.position;
        while self.ch.is_ascii_alphabetic() {
            self.read_char();
        }
        String::from_utf8_lossy(&self.input[start..self.position]).to_string()
    }

    fn read_number(&mut self) -> String {
        let start = self.position;
        while self.ch.is_ascii_digit() {
            self.read_char();
        }
        String::from_utf8_lossy(&self.input[start..self.position]).to_string()
    }
}

#[cfg(test)]
mod test {
    use anyhow::Result;

    use super::Lexer;
    use crate::token::Token;

    #[test]
    fn next_token() -> Result<()> {
        let input = "=+(){},;";
        let mut lexer = Lexer::new(input.into());

        let expected_tokens = vec![
            Token::Assign,
            Token::Plus,
            Token::LParen,
            Token::RParen,
            Token::LBrace,
            Token::RBrace,
            Token::Comma,
            Token::Semicolon,
        ];

        for token in expected_tokens {
            let t = lexer.next_token()?;
            assert_eq!(t, token);
        }
        Ok(())
    }

    #[test]
    fn int() -> Result<()> {
        let input = "let five = 5";
        let mut lexer = Lexer::new(input.into());
        let expected_tokens = vec![
            Token::Let,
            Token::Ident(String::from("five")),
            Token::Assign,
            Token::Int(String::from("5")),
        ];
        for token in expected_tokens {
            let t = lexer.next_token()?;
            assert_eq!(t, token);
        }
        Ok(())
    }

    #[test]
    fn next_complete() -> Result<()> {
        let input = r#"let five = 5;
            let ten = 10;
            let add = fn(x, y) {
                x + y;
            };
            let result = add(five, ten);
        !-/*5;
        5 < 10 > 5;
        if (5 < 10) {
            return true;
        } else {
            return false;
        }

        10 == 10;
        10 != 9;
        "#;

        let mut lex = Lexer::new(input.into());

        let tokens = vec![
            Token::Let,
            Token::Ident(String::from("five")),
            Token::Assign,
            Token::Int(String::from("5")),
            Token::Semicolon,
            Token::Let,
            Token::Ident(String::from("ten")),
            Token::Assign,
            Token::Int(String::from("10")),
            Token::Semicolon,
            Token::Let,
            Token::Ident(String::from("add")),
            Token::Assign,
            Token::Function,
            Token::LParen,
            Token::Ident(String::from("x")),
            Token::Comma,
            Token::Ident(String::from("y")),
            Token::RParen,
            Token::LBrace,
            Token::Ident(String::from("x")),
            Token::Plus,
            Token::Ident(String::from("y")),
            Token::Semicolon,
            Token::RBrace,
            Token::Semicolon,
            Token::Let,
            Token::Ident(String::from("result")),
            Token::Assign,
            Token::Ident(String::from("add")),
            Token::LParen,
            Token::Ident(String::from("five")),
            Token::Comma,
            Token::Ident(String::from("ten")),
            Token::RParen,
            Token::Semicolon,
            Token::Bang,
            Token::Minus,
            Token::Slash,
            Token::Asterisk,
            Token::Int(String::from("5")),
            Token::Semicolon,
            Token::Int(String::from("5")),
            Token::LessThan,
            Token::Int(String::from("10")),
            Token::GreaterThan,
            Token::Int(String::from("5")),
            Token::Semicolon,
            Token::If,
            Token::LParen,
            Token::Int(String::from("5")),
            Token::LessThan,
            Token::Int(String::from("10")),
            Token::RParen,
            Token::LBrace,
            Token::Return,
            Token::True,
            Token::Semicolon,
            Token::RBrace,
            Token::Else,
            Token::LBrace,
            Token::Return,
            Token::False,
            Token::Semicolon,
            Token::RBrace,
            Token::Int(String::from("10")),
            Token::Equal,
            Token::Int(String::from("10")),
            Token::Semicolon,
            Token::Int(String::from("10")),
            Token::NotEqual,
            Token::Int(String::from("9")),
            Token::Semicolon,
            Token::Eof,
        ];

        for token in tokens {
            let t = lex.next_token()?;
            assert_eq!(t, token);
        }

        Ok(())
    }
}
